### PR TITLE
Add service_msgs package

### PR DESCRIPTION
**Pull request by ihasdapie**
This PR is a prototype for the service introspection REP (https://github.com/ros-infrastructure/rep/pull/360) adding a `service_msgs` package implementing the proposed embedded message approach (https://github.com/ros-infrastructure/rep/pull/360#discussion_r902987531).

For instructions on how to build and run this see the meta-ticket https://github.com/ros2/ros2/issues/1285.

### CODE REVIEW

File: service_msgs/CHANGELOG.rst
Comment (ivanpauno) [inline service_msgs/CHANGELOG.rst]: this should be generated in the next release, you shouldn't add the file now

File: service_msgs/CMakeLists.txt
Comment (ivanpauno) [inline service_msgs/CMakeLists.txt]: May want to use:

```suggestion
  add_compile_options(-Wall -Wextra -Wconversion -Wsign-conversion -Wpedantic -Wnon-virtual-dtor -Woverloaded-virtual)
```

**Review by ivanpauno — COMMENTED** [review]
LGTM!

do you already have a PR for generating the new messages for each service?

File: service_msgs/CHANGELOG.rst
Comment (ihasdapie) [reply service_msgs/CHANGELOG.rst]: > LGTM!
> 
> do you already have a PR for generating the new messages for each service?

Yup, I'm just rebasing everything on jacob's rosidl changes before I make the PRs

Comment (jacobperron) [reply service_msgs/CHANGELOG.rst]: This file is newly added, but should be removed since it is typically generated at the time of a release.

File: service_msgs/CMakeLists.txt
Comment (jacobperron) [inline service_msgs/CMakeLists.txt]: Just adding a note here to remind us to update this to `rosidl_core_runtime` after https://github.com/ros2/rosidl_defaults/pull/22 is resolved.

File: service_msgs/QUALITY_DECLARATION.md
Comment (jacobperron) [inline service_msgs/QUALITY_DECLARATION.md]:
```suggestion
`service_msgs` has a list of provided [ROS interfaces](README.md).
```

File: service_msgs/QUALITY_DECLARATION.md
Comment (jacobperron) [inline service_msgs/QUALITY_DECLARATION.md]:
```suggestion
New interfaces require their own documentation in order to be added.
```

File: service_msgs/QUALITY_DECLARATION.md
Comment (jacobperron) [inline service_msgs/QUALITY_DECLARATION.md]: Commenting to remind us to update this line to `rosidl_core_runtime`.

File: service_msgs/QUALITY_DECLARATION.md
Comment (jacobperron) [inline service_msgs/QUALITY_DECLARATION.md]: We no longer run CI for mac
```suggestion
```

File: service_msgs/README.md
Comment (jacobperron) [inline service_msgs/README.md]:
```suggestion
This package contains message types used by ROS services.
```

File: service_msgs/README.md
Comment (jacobperron) [inline service_msgs/README.md]: This tutorial might be the best place to reference for an intro to ROS services: https://docs.ros.org/en/foxy/Tutorials/Beginner-CLI-Tools/Understanding-ROS2-Services/Understanding-ROS2-Services.html

File: service_msgs/README.md
Comment (jacobperron) [inline service_msgs/README.md]: I would elaborate here a bit more about how this is part of the service introspection feature (with a link to the REP).

File: service_msgs/msg/ServiceEventInfo.msg
Comment (jacobperron) [inline service_msgs/msg/ServiceEventInfo.msg]: Considering the discussion in https://github.com/ros2/rmw/pull/328, it's possible this identifier coming from the middleware may be more (or less) than 16 bytes (ie. it is not actually a UUID). So we probably want to update this to be an array of [24 bytes](https://github.com/ros2/rmw/blob/67b0f0a229c0bd665913a1d937ee82bf896ad554/rmw/include/rmw/types.h#L42), which is the current maximum and leave a comment where this is coming from.

**Comment by jacobperron** [comment]
rebased on rolling

File: service_msgs/CHANGELOG.rst
Comment (jacobperron) [reply service_msgs/CHANGELOG.rst]: Removed [4b8032a](https://github.com/ros2/rcl_interfaces/pull/143/commits/4b8032a1cf109abecd1fbec6e6e1818b235722d9)

File: service_msgs/CMakeLists.txt
Comment (jacobperron) [reply service_msgs/CMakeLists.txt]: Applied [283bf31](https://github.com/ros2/rcl_interfaces/pull/143/commits/283bf31fda7459882874f8dc54d3f8acb9dddb8f)

File: service_msgs/QUALITY_DECLARATION.md
Comment (jacobperron) [reply service_msgs/QUALITY_DECLARATION.md]: [8d94ae8](https://github.com/ros2/rcl_interfaces/pull/143/commits/8d94ae84e4ad1e533f0c7d57419024fca0ac1c08)

File: service_msgs/QUALITY_DECLARATION.md
Comment (jacobperron) [reply service_msgs/QUALITY_DECLARATION.md]: [8d94ae8](https://github.com/ros2/rcl_interfaces/pull/143/commits/8d94ae84e4ad1e533f0c7d57419024fca0ac1c08)

File: service_msgs/QUALITY_DECLARATION.md
Comment (jacobperron) [reply service_msgs/QUALITY_DECLARATION.md]: [8d94ae8](https://github.com/ros2/rcl_interfaces/pull/143/commits/8d94ae84e4ad1e533f0c7d57419024fca0ac1c08)

File: service_msgs/QUALITY_DECLARATION.md
Comment (jacobperron) [reply service_msgs/QUALITY_DECLARATION.md]: [8d94ae8](https://github.com/ros2/rcl_interfaces/pull/143/commits/8d94ae84e4ad1e533f0c7d57419024fca0ac1c08)

File: service_msgs/README.md
Comment (jacobperron) [reply service_msgs/README.md]: [f274555](https://github.com/ros2/rcl_interfaces/pull/143/commits/f27455529315fbbc6f49a16f852156e379bab2c5)

File: service_msgs/README.md
Comment (jacobperron) [reply service_msgs/README.md]: [f274555](https://github.com/ros2/rcl_interfaces/pull/143/commits/f27455529315fbbc6f49a16f852156e379bab2c5)

File: service_msgs/README.md
Comment (jacobperron) [reply service_msgs/README.md]: [f274555](https://github.com/ros2/rcl_interfaces/pull/143/commits/f27455529315fbbc6f49a16f852156e379bab2c5)

File: service_msgs/CMakeLists.txt
Comment (jacobperron) [reply service_msgs/CMakeLists.txt]: This was resolved in [f274555](https://github.com/ros2/rcl_interfaces/pull/143/commits/f27455529315fbbc6f49a16f852156e379bab2c5)

Comment (jacobperron) [reply service_msgs/CMakeLists.txt]: and [39827df](https://github.com/ros2/rcl_interfaces/pull/143/commits/39827df73c2bc21ce97e14856a97dc4890a65eab)

File: service_msgs/msg/ServiceEventInfo.msg
Comment (ivanpauno) [reply service_msgs/msg/ServiceEventInfo.msg]: I think all implementations use 16 bytes (128 bits).
I'm not sure why 24 bytes was chosen in rmw.

**Review by ivanpauno — APPROVED** [review]
LGTM, we should double check if using `unique_identifier_msgs/UUID` is fine or not before merging

**Comment by ivanpauno** [comment]
Rebased and added https://github.com/ros2/rcl_interfaces/pull/143/commits/d0c30965814995806d2c1f460e98b395d5201698, if not `action_msgs` doesn't build

File: service_msgs/msg/ServiceEventInfo.msg
Comment (wjwwood) [reply service_msgs/msg/ServiceEventInfo.msg]: Jacob and I discussed this before he left for his travels, but regardless of the size of this field, using UUID for it, is not correct, in my opinion, if the `client_id` is the entity ID from DDS/rmw. They are GID/GUID and are not guaranteed to be "universally" unique like a UUID.

So we should use a different type for this.

However, if I'm wrong and `client_id` is something we're generating at the ROS level and it is a UUID this is ok, but I don't think that's the case off-hand.

Comment (ivanpauno) [reply service_msgs/msg/ServiceEventInfo.msg]: Sounds good!

Should I create a new message type called GID? is that name fine?
Where should I create it? is builtin_interfaces a good place? in a new package?
Do we need a new message type? Maybe I can use an array directly in `ServiceEventInfo`.

**Comment by ivanpauno** [comment]
CI:

* Linux [Build Status](http://ci.ros2.org/job/ci_linux/17863/)
* Linux-aarch64 [Build Status](http://ci.ros2.org/job/ci_linux-aarch64/12431/)
* Windows [Build Status](http://ci.ros2.org/job/ci_windows/18534/)

File: service_msgs/msg/ServiceEventInfo.msg
Comment (clalancette) [reply service_msgs/msg/ServiceEventInfo.msg]: After a lot of discussion, we did in fact come to conclusion that this should be a GID (as defined in the DDS spec), rather than a UUID.  They both happen to be 16-bytes long, but they are semantically different (and DDS assigns certain meaning to certain bytes).

I attempted to reuse the GID message from https://github.com/ros2/rmw_dds_common/blob/rolling/rmw_dds_common/msg/Gid.msg , but I ran into circular dependency issues between this repository and rmw_dds_common.  So what I ended up doing here was to call the field client_gid, and make it a 16-byte char field.  This also required a few other changes in some other repositories, and that is now all pushed.

So I think we can call this particular conversation resolved.  I'm going to mark it as such, kick off CI on the whole lot of PRs, and see where we are.

**Review by gbiggs — APPROVED** [review]

**Review by mjcarroll — APPROVED** [review]

**Comment by clalancette** [comment]
CI for this is in https://github.com/ros2/ros2/issues/1285#issuecomment-1404070715

**Pull request merged**
